### DISCUSSION TITLE
`io_uring_prep_send_zc()` performing a little worst than `io_uring_prep_send()`

### DISCUSSION THREAD

**Discussion by jmillan** [axboe/liburing#1003] [Q&A] [answered]
I'm trying out `io_uring_prep_send_zc()` and comparing it with `io_uring_prep_send()` in my server.

In order to enable zero copy I do as follows:

#### Enable SO_ZEROCOPY in the socket file descriptor

```c++
// Enable zero-copy in the file descriptor.
int on = 1;

err = setsockopt(this->fd, SOL_SOCKET, SO_ZEROCOPY, &on, sizeof(on));
```

#### Register buffers into io_uring
```c++
 // Register buffers.
for (size_t i{ 0 }; i < DepLibUring::QueueDepth; ++i)
{
 this->iov[i].iov_base = malloc(1500);
 this->iov[i].iov_len = 1500;
}

err = io_uring_register_buffers(std::addressof(this->ring), this->iov, DepLibUring::QueueDepth);
```

### Prepare send
```c++
io_uring_prep_send_zc(sqe, sockfd, this->iov[idx].iov_base,  this->iov[idx].iov_len, 0, 0);
io_uring_prep_send_set_addr(sqe, addr, addrlen);

sqe->ioprio |= IORING_RECVSEND_FIXED_BUF;
sqe->buf_index = idx;
```

I'm handling notifications accordingly, making the buffers available when receiving the `IORING_CQE_F_NOTIF` in `cqe-flags`.

Tracing the process with and without zero copy I see that there is a slightly more CPU usage when using `ZC`. These are some observations comparing the traces:

* There is a function `io_import_fixed` being called, taking a little time. That is no present in `NO_ZC`. It's expected.
* There is a function `io_send_zc_prep` being called, takes a little more time than the analogous `io_sendmsg_prep`.
* `ip_append_data` takes considerably less time in `ZC` in comparison with `NO_ZC`. I can imagine this is where `CZ` shines.
* `ip_send_skb` function:
   *   `__ip_local_out` seems to take the same time in both cases.
   * `ip_finish_output` takes more time in `ZC` than in `NO_ZC`. 

From these observations I see that the benefit of using `ZC` is in `ip_append_data`, but for some reason it takes more time than `NO_ZC` in the other mentioned functions, so in overall there is no benefit.

In my tests I'm sending around 200K UDP (audio and video) packets through a socket.

I would definitely like to understand this numbers better, as everything pointed that `ZC` would show improvements compared to `NO_ZC`.

Here the most significant part of the traces flamegraph for a visual comparison:

#### NO_ZC
<img width="1200" alt="Screenshot 2023-11-23 at 16 14 07" src="https://github.com/axboe/liburing/assets/732340/53048c97-2727-4f62-a03f-d8db42f96f8c">

### ZC
<img width="1193" alt="Screenshot 2023-11-23 at 16 16 43" src="https://github.com/axboe/liburing/assets/732340/e9d258af-3864-4f6a-9c26-dbd34157fef9">

**Comment by jmillan** [comment]
NOTE: I can provide the flamegraph .svg files in case there is any interest.

**Comment by isilence** [comment]
> In order to enable zero copy I do as follows:
> #### Enable SO_ZEROCOPY in the socket file descriptor
> 
> ```c++
> // Enable zero-copy in the file descriptor.
> int on = 1;
> 
> err = setsockopt(this->fd, SOL_SOCKET, SO_ZEROCOPY, &on, sizeof(on));
> ```

`SO_ZEROCOPY` is not needed for io_uring zc and even harmful. Just a general note, it wouldn't change performance.

> I'm handling notifications accordingly, making the buffers available when receiving the `IORING_CQE_F_NOTIF` in `cqe-flags`.
> 
> Tracing the process with and without zero copy I see that there is a slightly more CPU usage when using `ZC`. These are some observations comparing the traces:

That's how it often goes, CPUs are pretty good at copying small buffers, assuming it's <=1500B per request in your test, and zerocopy also adds some additional overhead.  The kernel's networking stack is also pretty heavy in general, saving on small copies wouldn't make a drastic change. You might try to enable gso (e.g. `UDP_SEGMENT`), or if possible bump up the request size (together with network mtu).

>     * There is a function `io_import_fixed` being called, taking a little time. That is no present in `NO_ZC`. It's expected.
> 
>     * There is a function `io_send_zc_prep` being called, takes a little more time than the analogous `io_sendmsg_prep`.
> 
>     * `ip_append_data` takes considerably less time in `ZC` in comparison with `NO_ZC`. I can imagine this is where `CZ` shines.

Yes, that's where it copies data. 

>     * `ip_send_skb` function:
>       
>       * `__ip_local_out` seems to take the same time in both cases.
>       * `ip_finish_output` takes more time in `ZC` than in `NO_ZC`.

That's interesting, I don't see immediately why. Can you share the svg of the profiles as you offered?

**Reply by jmillan** [reply]
> SO_ZEROCOPY is not needed for io_uring zc and even harmful. Just a general note, it wouldn't change performance.

Thanks for the info 👍 .

> You might try to enable gso (e.g. UDP_SEGMENT),

I'll take a look at this.

> or if possible bump up the request size (together with network mtu).

This is a constraint for the service, there are no packets sent or received >1500B.

> Can you share the svg of the profiles as you offered?

Of course, yes.

EDIT: In order to see properly visualize the .svg's you may need to download and open locally.

#### NO_ZC
![perf_40-AV-uring_nzc](https://github.com/axboe/liburing/assets/732340/fbce771a-426c-4e2e-b79f-1a77a7cccdd0)

#### ZC
![perf_40-AV-uring_zc](https://github.com/axboe/liburing/assets/732340/a4aa6d69-bf91-4ee4-94fc-008f864d0ddd)

**Reply by jmillan** [reply]
After reading about UDP_SEGMENT, it's something that does not apply to our server, nor the MTU increase.

If I may briefly explain the nature of this server, it's basically a RTP (Real Time Protocol) relay that receives traffic (ie: audio, video) from one socket and relays such traffic to N other sockets. Each relayed packet is slightly modified, as we encrypt the data for each destination based on the uniquely selected keys.

In a nutshell, the nature of this traffic is real time media, the traffic received comes out of capturing a media device (microphone, webcam, etc), encoding the media, packaging the content into RTP packets and sending it to the server who will deliver such traffic to N endpoints that will un-package the data, extract the media and play it out of the corresponding media device (speaker, screen, etc).

This been said, each network packet needs to be delivered as it comes, so using UDP_SEGMENT is not an option as we can't wait to receive multiple packets in order to deliver them.

liburing has so far enhanced the performance by reducing the system calls for `sendto`. Since we relay one packet to N destinations, liburing allows us to `prepare` all sending packets and effectively send them within a single call to `io_uring_submit()`.

I thought `ZC` was also going to help here, but as shown it does not. I still can't see how `ip_finish_output2` takes more time with `ZC` than without it. Do you have any idea on how could this be further debugged?

**Reply by isilence** [reply]
First to note, we're looking at ~0.5% differences in functions down the stack, so it might be naturally noisy and hard to capture precise ratios. I also see that coredumping takes ~0.5%, which wouldn't usually show up in long enough testing. Let me ask, for how long you was sampling it? Would be great to have >1 min runs.

With that aside, let's zoom into `start_xmit` -- 6.86% (non-zc) vs 7.43% (zc), which is 0.57% difference, and let's assume it's a real difference, but not normalization, noise and other side effects.

1) `virtqueue_add_outbuf()` does extra allocations in case of zerocopy and costs extra 0.26%. I didn't check what's happening in virtqueue, but the most likely reason is that for normal sends the kernel will allocate a buffer and try to pack all headers and user data in it, so we pass only 1 page to virtqueue per send. For zerocopy, headers will still stay in separately allocated in kernel buffer, but user data would point to a user page, so we have 2 pages there per request. And looks it's matched with a `free()` in `virtqueue_get_buf_ctx_split()`.

2) Now, I mentioned zerocopy brings some extra overhead, one of them is those notification completions. We can see them in
`... -> start_xmit() -> free_old_xmit_skbs() -> ... -> skb_release_data() -> io_uring_zerocopy_callback() -> __io_req_task_work_add()` 
There is not much I can do about that, most of it is adding ourselves into a list (1 cmpxchg when unconteded). The first version of the zerocopy API was a bit more complicated but allowed having a single notification CQEs per multiple sends. I might return it back if there is a really good case for that, but it's only 0.1% here.

It'll also try to wake up the task, but you're lucky here because it already runs in the task context and so doesn't have to actually wake it up, which is pretty expensive. It's usually not like that with other drivers, especially for real hw. In fact, the trace is only for the task but doesn't cover irqs, which might be calling `io_uring_zerocopy_callback` and adding overhead that we don't see.

3) Another part of it is in `do_syscall_64() -> syscall_exit_to_user_mode() -> ... -> task_work_run()`, ~0.25%. 

There is a feature that can help a bit with 2) and 3) called `IORING_SETUP_DEFER_TASKRUN`, but it should go together with `IORING_SETUP_SINGLE_ISSUER` and adds some limitations on the userspace, i.e. only one task can use the ring, but seems the test/app is single threaded anyway. The user is also has to do a syscall to get completions (e.g. via io_uring_wait_cqes()).

Also several notes not related to zerocopy:

- You may try registering the ring fd, see `IORING_ENTER_REGISTERED_RING` and `IORING_REGISTER_RING_FDS`.
- Try out `IORING_REGISTER_FILES`
- Tying io_uring into epoll via eventfd incurs overhead.
- You can try to convert recvmsg(2) to io_uring, and take a look at multishot recv.

**Reply by jmillan** [reply, marked as answer]
I've repeated the test to be longer than 1 minute order to remove the possible noises.

NOTES:
* IORING_REGISTER_FILES is used now.
* IORING_REGISTER_RING_FDS not used as I got `Operation not supported` as response to `io_uring_register_ring_fd()`.
  *  kernel version: 6.1.0-13 
* Encryption has been disabled now, it's a legitimate case and was providing too much noise on the traces.

As you can see there is now a 2% difference between `ZC` and `NO_ZC`, being `ZC` more efficient.

### ZC
![perf_40-AV-uring_zc_NEW](https://github.com/axboe/liburing/assets/732340/5b784b61-9eed-4d6a-b24d-aebe13d4f2cb)

### NO ZC
![perf_40-AV-uring_no_zc_NEW](https://github.com/axboe/liburing/assets/732340/2d1bdd71-c5ae-485e-805f-82f28ce61705)


Probably now we are seeing the benefits of using `ZC` compared to not using it, for our use case (small buffers). I hope everything makes sense.

>Also several notes not related to zerocopy:

>You may try registering the ring fd, see IORING_ENTER_REGISTERED_RING and IORING_REGISTER_RING_FDS.

Tried, but got the mentioned error. Maybe it's only enabled for newer kernels.

>Try out IORING_REGISTER_FILES

New traces do contain it.

>Tying io_uring into epoll via eventfd incurs overhead.

This is a convenience as our server lies on top of [libuv](https://libuv.org/) event loop. The server does not handle threads. I'm registering an eventfd to the io_uring and letting libuv pool such eventfd and notify me whenever there are CQEs.

>You can try to convert recvmsg(2) to io_uring, and take a look at multishot recv.

We are actually using `recvmmsg`. Maybe we could take advantage of io_uring multishot recv, but I was trying the sending side as the server relies N packets for each one it receives, so the highest impact can happen in the sending side. Also our server works for other OOSS apart from Linux, so I'd rather keep using `libuv` for reception.

**Reply by isilence** [reply]
> As you can see there is now a 2% difference between ZC and NO_ZC, being ZC more efficient.

Sweet, glad to see that.

> Probably now we are seeing the benefits of using ZC compared to not using it, for our use case (small buffers). I hope everything makes sense.

Getting rid of virtio allocations might be a pretty easy win here. From a quick look it seems because of this line:

https://elixir.bootlin.com/linux/v6.7-rc3/source/drivers/virtio/virtio_ring.c#L251

I'd bump the threshold to 2 and see if it helps.

```C
diff --git a/drivers/virtio/virtio_ring.c b/drivers/virtio/virtio_ring.c
index 51d8f3299c10..d6170cbd9e49 100644
--- a/drivers/virtio/virtio_ring.c
+++ b/drivers/virtio/virtio_ring.c
@@ -248,7 +248,7 @@ static bool virtqueue_use_indirect(const struct vring_virtqueue *vq,
 	 * If the host supports indirect descriptor tables, and we have multiple
 	 * buffers, then go indirect. FIXME: tune this threshold
 	 */
-	return (vq->indirect && total_sg > 1 && vq->vq.num_free);
+	return (vq->indirect && total_sg > 2 && vq->vq.num_free);
 }
 
 /*
```

**Reply by jmillan** [reply]
> I'd bump the threshold to 2 and see if it helps.

I'll try this when possible and report the results :+1:

**Reply by jmillan** [reply]
@isilence,

Appologizes.., It's been too long since I did compile a Linux kernel. I imagine in this case only `virtio` module needs to be compiled. Would you point me somewhere with the doc to compile a module out of the recent Linux sources? 

Another question: how comes `virtio` is being used? Is it due to Linux being virtualized in this case, or does `io_uring` use it on every case?

Thanks

**Reply by isilence** [reply]
> how comes virtio is being used? Is it due to Linux being virtualized in this case, or does io_uring use it on every case?

It's because it's a VM and it's configured to use virtio for net, unrelated to io_uring. And I forgot to mention, that last diff won't give you anything, if you use VMs only for testing, but it's still would be useful in general to fix it up if numbers back the diff.

**Reply by isilence** [reply]
> Appologizes.., It's been too long since I did compile a Linux kernel. I imagine in this case only virtio module needs to be compiled.

It'll be easier to compile it all, not just one module.  

> Would you point me somewhere with the doc to compile a module out of the recent Linux sources?

https://wiki.archlinux.org/title/Kernel/Traditional_compilation

Try this one, that's how it usually looks. It's for Arch but should be fine.
Instead of step 1. you can pick any repository, e.g.

```
git://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git
```

After checkout tag `v6.6`, and go to step 2. and onwards (4.4 can be skipped).

Note that it assumes you're doing it all on the target machine/VM and installing it onto a drive (e.g. a QEMU virtual drive or so). There is also an option to compile it outside and pass a kernel image to qemu,

**Reply by jmillan** [reply]
These are the results with the modified `virtio_ring` module. At a first glance I see no real difference. 

NOTE1: I've installed the module on the virtual linux machine where I'm running the tests.

              EDIT: This is, I only touched the virtual linux image itself. This is where I upgrade debian, etc. So considering it's the 
                         only that is using this kernel IMHO this must be the place where `virtio_ring` must be modified. I guess
                         this is what `target machine` means.
NOTE2: I used the kernel version I use in the server currently: 6.1.0-13

### ZC with virtio_ring changes
![perf_40-AV-uring_zc_virtio_ring](https://github.com/axboe/liburing/assets/732340/163e7a3a-fb8f-4238-af91-47dc22955ad8)

**Reply by jmillan** [reply]
I think I see no improvement nor the opposite.

**Reply by isilence** [reply]
And the allocation is still there, the diff didn't work out. Well, that means I need to try it  myself  sometime later and see how to force the non-allocation path.

**Reply by jmillan** [reply]
Thanks a lot for all your comments @isilence. I've marked this question as answered.

**Reply by jmillan** [reply]
> It's because it's a VM and it's configured to use virtio for net, unrelated to io_uring. And I forgot to mention, that last diff won't give you anything, if you use VMs only for testing, but it's still would be useful in general to fix it up if numbers back the diff.

Definitely this is a service that is highly run in VM.

If you want me to test some recent kernel in the future let me know so I can bench and give the .svg's back for analising.

**Reply by jmillan** [reply]
Woah.., I was giving it a try to monitor a Docker image.., this is how it looks the trace.., there are so many stacked calls compared to previous ones.

I'll check the `ZC` benefits in this scenario.

![docker3](https://github.com/axboe/liburing/assets/732340/d35fef84-e62c-4e36-92c6-cdf7bb0495ba)

**Reply by isilence** [reply]
> Woah.., I was giving it a try to monitor a Docker image.., this is how it looks the trace.., there are so many stacked calls compared to previous ones.

So it's a **veth** interface pair that connects a container to the outside. I also seems it's bridging it to a "real" network interface. In quotes because judging from virtio at the end, it's all in a VM?

All those additional traces you see are the veth receive path, i.e. it's sending from one end of veth, then the other veth end receives it and doing sth else like forwarding. It's not super efficient in the first place, but also all such places where sends turn into recv inside a single machine will explicitly disable zc send. So, it'll likely regress with zc instead of gaining anything unless you forward traffic directly

edit: I probably screwed terminology here. Also "within a single machine" should be "within a single kernel" as it doesn't really concern VMs.

**Reply by jmillan** [reply]
> > Woah.., I was giving it a try to monitor a Docker image.., this is how it looks the trace.., there are so many stacked calls compared to previous ones.
> 
> So it's a **veth** interface pair that connects a container to the outside. I also seems it's bridging it to a "real" network interface. In quotes because judging from virtio at the end, it's all in a VM?
> 

In this case the concrete scenario is an Ubuntu docker image running on a MacOS (using Docker Desktop).

The previous traces were taken from a Linux container using kvm.

So the ZC gain will probably be  more evident when running Linux directly on the machine?

**Reply by isilence** [reply]
The problem is in containers, or to be more precise connecting them via (default) veth. That middle part should look pretty similar even if you'd be running Linux bare metal.

I'm not familiar enough to say what would be a better option, but you might find interesting looking through this talk/slides:
https://lpc.events/event/17/contributions/1581/

**Reply by isilence** [reply]
```C
diff --git a/drivers/virtio/virtio_ring.c b/drivers/virtio/virtio_ring.c
index 51d8f3299c10..d6170cbd9e49 100644
--- a/drivers/virtio/virtio_ring.c
+++ b/drivers/virtio/virtio_ring.c
@@ -248,7 +248,7 @@ static bool virtqueue_use_indirect(const struct vring_virtqueue *vq,
 	 * If the host supports indirect descriptor tables, and we have multiple
 	 * buffers, then go indirect. FIXME: tune this threshold
 	 */
-	return (vq->indirect && total_sg > 1 && vq->vq.num_free);
+	return (vq->indirect && total_sg > 2 && vq->vq.num_free);
 }
 
 /*
```

I've tried it out, at least for me that works. Do you mind rerunning it? My first guess is that maybe there are more than 2 pages. Can you remind what the send size and whether the buffers are page aligned?

```
bpftrace -e "kprobe:virtqueue_add_outbuf { print(arg3); } " -- ./program
```

Not tested, but this tracer should show exactly what we want, unpatched kernel is fine. You can also run bpftrace separately and specify pid and/or save the output into a file with `-o`. 

The second guess is that maybe there was something wrong with installing a patched kernel, did you check what `uname -r` was saying?

**Reply by jmillan** [reply]
Running the command: `sudo bpftrace -e "kprobe:virtqueue_add_outbuf /pid == 1442/ { print(arg3); } "` (being 1442 the PID), I'm getting a lot of output such as:

```txt
18446629531275940096
18446629534034071296
18446629531275943680
18446629531321990144
18446629531275940352
18446629531321989376
18446629531237027072
18446629531275939840
18446629532127728384
18446629531275941888
18446629531275941376
18446629531275942656
18446629532127730432
18446629532127728128
18446629532127728640
18446629533465643008
18446629532127728896
18446629533465643776
18446629533465641216
18446629533465641984
18446629532127730688
18446629531321989632
...
```

> Can you remind what the send size and whether the buffers are page aligned?

The buffers passed to `iov` are 1500B big.  I can't say about the alignment, all the buffer is allocated within a single call to malloc, this is, 1500*N, being N the iov count.

> The second guess is that maybe there was something wrong with installing a patched kernel, did you check what uname -r was saying?

I just built the modified module and substituted the binary of the previous one by the newest in the corresponding path.

**Reply by isilence** [reply]
```
18446629531275940096
18446629534034071296
...
```

That's not right... Can you try arg2?

```
bpftrace -e "kprobe:virtqueue_add_outbuf { print(arg2); } "
```

**Reply by jmillan** [reply]
> That's not right... Can you try arg2?

In this case what I see is '1'.

Meaning the patch didn't apply, right?

**Reply by isilence** [reply]
Can you share a profile of that run? With zc I expect it to be 2-3, which is why there were extra allocations in `virtqueue_add_split`. And for non-zerocopy it should likely be 1 and without any allocations.

**Reply by jmillan** [reply]
I'l try to do it in the next days.

What do you need, the output file of `perf record`?

**Reply by isilence** [reply]
Thanks. perf should be enough, `perf  report` or a flamegraph. Just set the threshold limit in case the allocs are under 0.5%.

```
perf report --call-graph=graph,0
```

**Reply by jmillan** [reply]
Here the requested info @isilence:

## ZC

Result of `bpftrace -e "kprobe:virtqueue_add_outbuf { print(arg2); } "` is `2 and 3`.

![perf_zc](https://github.com/axboe/liburing/assets/732340/208c18f7-a37b-4578-b58e-53fda2884d3e)


### NO_ZC

Result of `bpftrace -e "kprobe:virtqueue_add_outbuf { print(arg2); } "` is `1`.

![perf_no_zc](https://github.com/axboe/liburing/assets/732340/5ee089e6-fafa-4dca-a14d-8fa204e0cff0)

**Reply by isilence** [reply]
That's good, it behaves as expected. What I ultimately need is performance numbers, would you mind benchmarking it for me?

I'd ask to make it so buffers don't span across multiple pages. Considering 1500B, it can be 2 buffers per page, or just make it one buffer per page if you have enough memory. Also, would also be great to zoom in on the sendmsg overhead, e.g. by disabling encryption like last time. Maybe removing rx as well if there is already an option for that.

I need to compare zerocopy with and without the patch. If the CPU is saturated and running at 100%, then I'm looking for difference in qps. If it's not, we'd need to look for how CPU util is changed. I don't remember which one is true for your testing. Also, and let it run for >5 minutes each, hopefully with a short prior warmup.

**Reply by jmillan** [reply]
Hi @isilence, since I'm not a kernel insider you'll have to explain to me a bit more :-)

> I'd ask to make it so buffers don't span across multiple pages. Considering 1500B, it can be 2 buffers per page, or just make it one buffer per page if you have enough memory. 

What does this really mean? How can I do that?

> Also, would also be great to zoom in on the sendmsg overhead, e.g. by disabling encryption like last time.

:+1:

> If the CPU is saturated and running at 100%, then I'm looking for difference in qps. If it's not, we'd need to look for how CPU util is changed. I don't remember which one is true for your testing.

As per top/htop, the CPU never goes 100%, indeed it remains in 13% aprox.

> we'd need to look for how CPU util is changed.

What does this mean?

> let it run for >5 minutes each, hopefully with a short prior warmup.

:+1:

NOTE: Just so you can have a sense of how we're using liburing, this is the code, which resides in just a .cpp and .hpp file: https://github.com/versatica/mediasoup/blob/v3/worker/src/DepLibUring.cpp#L322

**Reply by isilence** [reply]
> Hi @isilence, since I'm not a kernel insider you'll have to explain to me a bit more :-)

Absolutely!

> > I'd ask to make it so buffers don't span across multiple pages. Considering 1500B, it can be 2 buffers per page, or just make it one buffer per page if you have enough memory.
> 
> What does this really mean?

Oversimplifying, let's say a page is a block of memory of size of 4096 bytes and also aligned to 4096, i.e. `(unsigned long)p % 4096 == 0`. A buffer is in a single page if it fully falls into one of such blocks.

> How can I do that?

Looks like your buffers come like this:

```C++
using SendBuffer = uint8_t[SendBufferSize];
SendBuffer sendBuffers[QueueDepth];
```

In which case you can just replace with `SendBuffer` with the struct below and let the compiler take care of all alignment for you (once you fix up compilation errors).

```C++
struct SendBuffer {
    char data[1500];
} __attribute__ ((aligned (4096)));
```

It's for benchmarking only, that  aligned attribute here will waste lots of memory, as it'd grow the SendBuffer size to 4096, while you only use 1500 bytes of it. 

> > we'd need to look for how CPU util is changed.
> 
> What does this mean?

Looking through the thread, you already was measuring CPU usage before, it's the same thing

**Reply by jmillan** [reply]
Hi,

Surprisingly, aligned structure takes more time than no aligned one.

* Samples were taken during 5 minutes each.
* Using your patch for the kernel.
* Encryption was disabled to avoid noise.

### Aligned data structure (4096)
![40_liburing_aligned](https://github.com/axboe/liburing/assets/732340/cd4276d8-1538-439f-b05f-797e6c618301)

### Unaligned data structure
![40_liburing_no_aligned](https://github.com/axboe/liburing/assets/732340/2f17a05b-b208-4c26-ac6d-2e6478eef5df)

**Reply by isilence** [reply]
Thanks @jmillan, I was rather asking to compare with and without the patch, but looking and the profiles it looks like the optimisation didn't work out for you, weird. 

> Surprisingly, aligned structure takes more time than no aligned one.

Not unprecedented, it may be due to increased pressure on TLB or because caching and prefetching is worse with gaps. Or it might be just variance b/w runs (fluctuating CPU freq and so on). Good for benchmarking, but wouldn't recommend for prod because it wastes too much memory.
